specs/core/0_beacon-chain.md

```diff
@@ -1301,128 +1283,99 @@ def process_deposit(state: BeaconState,
 
 ```python
-def activate_validator(state: BeaconState,
-                       index: int) -> None:
-    """
-    Activate the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
+def activate_validator(state: BeaconState, index: int, genesis: bool) -> None:
     validator = state.validator_registry[index]
-    if validator.status != PENDING_ACTIVATION:
-        return
 
-    validator.status = ACTIVE
-    validator.latest_status_change_slot = state.slot
-    state.validator_registry_delta_chain_tip = get_new_validator_registry_delta_chain_tip(
-        current_validator_registry_delta_chain_tip=state.validator_registry_delta_chain_tip,
-        validator_index=index,
-        pubkey=validator.pubkey,
-        flag=ACTIVATION,
+    validator.activation_slot = GENESIS_SLOT if genesis else (state.slot + ENTRY_EXIT_DELAY)
+    state.validator_registry_delta_chain_tip = hash_tree_root(
+        ValidatorRegistryDeltaBlock(
+            current_validator_registry_delta_chain_tip=state.validator_registry_delta_chain_tip,
+            validator_index=index,
+            pubkey=validator.pubkey,
+            slot=validator.activation_slot,
+            flag=ACTIVATION,
+        )
     )
 ```
 
 ```python
-def initiate_validator_exit(state: BeaconState,
-                            index: int) -> None:
-    """
-    Initiate exit for the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
+def initiate_validator_exit(state: BeaconState, index: int) -> None:
     validator = state.validator_registry[index]
-    if validator.status != ACTIVE:
-        return
-
-    validator.status = ACTIVE_PENDING_EXIT
-    validator.latest_status_change_slot = state.slot
+    validator.status_flags |= INITIATED_EXIT
 ```
 
 ```python
-def exit_validator(state: BeaconState,
-                   index: int,
-                   new_status: int) -> None:
-    """
-    Exit the validator with the given ``index``.
-    Note that this function mutates ``state``.
-    """
+def exit_validator(state: BeaconState, index: int) -> None:
     validator = state.validator_registry[index]
-    prev_status = validator.status
 
-    if prev_status == EXITED_WITH_PENALTY:
-        return
-
-    validator.status = new_status
-    validator.latest_status_change_slot = state.slot
-
-    if new_status == EXITED_WITH_PENALTY:
-        state.latest_penalized_exit_balances[state.slot // COLLECTIVE_PENALTY_CALCULATION_PERIOD] += get_effective_balance(state, index)
-
-        whistleblower_index = get_beacon_proposer_index(state, state.slot)
-        whistleblower_reward = get_effective_balance(state, index) // WHISTLEBLOWER_REWARD_QUOTIENT
-        state.validator_balances[whistleblower_index] += whistleblower_reward
-        state.validator_balances[index] -= whistleblower_reward
-
-    if prev_status == EXITED_WITHOUT_PENALTY:
+    if validator.exit_slot < state.slot + ENTRY_EXIT_DELAY:
```

Comment:
It's not actually possible to push the exit event back further, because the exit slot is the current slot plus 256, so later exits will have later exit slots, so this line will throw them out.

More philosophically, the principle I'm going after here is "state changes only succeed if they bring the state change slot _earlier_". If hypothetically we add a feature where voluntary exits take 3 weeks but penalized exits happen immediately, then it would work as is as I wrote it, but not if we change it to FAR_FUTURE_SLOT.

Same applies to the other similar cases below.